scipy/special/cephes/psi.c

```diff
@@ -67,73 +76,122 @@ static double A[] = {
     8.33333333333333333333E-2
 };
 
-double psi(x)
-double x;
+
+double digamma_imp_1_2(double x)
 {
-    double p, q, nz, s, w, y, z;
-    int i, n, negative;
+    /*
+     * Rational approximation on [1, 2] taken from Boost.
+     *
+     * Now for the approximation, we use the form:
+     *
+     * digamma(x) = (x - root) * (Y + R(x-1))
+     *
+     * Where root is the location of the positive root of digamma,
+     * Y is a constant, and R is optimised for low absolute error
+     * compared to Y.
+     *
+     * Maximum Deviation Found:               1.466e-18
+     * At double precision, max error found:  2.452e-17
+     */
+    double r, g;
+
+    static const float Y = 0.99558162689208984f;
+
+    static const double root1 = 1569415565.0 / 1073741824.0;
+    static const double root2 = (381566830.0 / 1073741824.0) / 1073741824.0;
+    static const double root3 = 0.9016312093258695918615325266959189453125e-19;
+
+   static double P[] = {
+       -0.0020713321167745952,
+       -0.045251321448739056,
+       -0.28919126444774784,
+       -0.65031853770896507,
+       -0.32555031186804491,
+       0.25479851061131551
+   };
+   static double Q[] = {
+       -0.55789841321675513e-6,
+       0.0021284987017821144,
+       0.054151797245674225,
+       0.43593529692665969,
+       1.4606242909763515,
+       2.0767117023730469,
+       1.0
+   };
+   g = x - root1;
+   g -= root2;
+   g -= root3;
+   r = polevl(x - 1.0, P, 5) / polevl(x - 1.0, Q, 6);
+
+   return g * Y + g * r;
+}
+
+
+double psi_asy(double x)
+{
+    double y, z;
+
+    if (x < 1.0e17) {
+	z = 1.0 / (x * x);
+	y = z * polevl(z, A, 6);
+    }
+    else {
+	y = 0.0;
+    }
 
-    negative = 0;
-    nz = 0.0;
+    return log(x) - (0.5 / x) - y;
+}
 
-    if (x <= 0.0) {
-	negative = 1;
-	q = x;
-	p = floor(q);
-	if (p == q) {
+
+double psi(double x)
+{
+    double y = 0.0;
+    double q, r, w;
+    int i, n;
+
+    if (x == 0) {
+	mtherr("psi", SING);
+	return npy_copysign(NPY_INFINITY, -x);
+    }
+    else if (x < 0.0) {
+	/* argument reduction before evaluating tan(pi * x) */
+	r = modf(x, &q);
+	if (r == 0.0) {
 	    mtherr("psi", SING);
-	    return (NPY_INFINITY);
-	}
-	/* Remove the zeros of tan(NPY_PI x)
-	 * by subtracting the nearest integer from x
-	 */
-	nz = q - p;
-	if (nz != 0.5) {
-	    if (nz > 0.5) {
-		p += 1.0;
-		nz = q - p;
-	    }
-	    nz = NPY_PI / tan(NPY_PI * nz);
-	}
-	else {
-	    nz = 0.0;
+	    return NPY_NAN;
 	}
+	y = -NPY_PI / tan(NPY_PI * r);
 	x = 1.0 - x;
     }
 
     /* check for positive integer up to 10 */
     if ((x <= 10.0) && (x == floor(x))) {
-	y = 0.0;
 	n = x;
```

Comment:
Re: removal of `y=0` initialization --- what happens when `x<0` and `1 - x` rounds to `1.0`?
Was the old code wrong?

Reply:
Huh, good catch. The old code was wrong, and so is the new code.

```python
>>> import scipy.special as sc
>>> import mpmath
>>> sc.digamma(-1e-17)
99999999999999984.0
>>> mpmath.digamma(-1e-17)
mpf('99176043407040600.0')
>>> import scipy; scipy.__version__
'1.0.0'
```

```python
>>> import scipy.special as sc
>>> import mpmath
>>> sc.digamma(-1e-17)
99999999999999984.0
>>> mpmath.digamma(-1e-17)
mpf('99176043407040600.0')
>>> import scipy; scipy.__version__
'1.1.0.dev0+233ce70'
```

Reply:
I take it back, just needed to use more precision in Mpmath:

```python
>>> with mpmath.workdps(100):
...     mpmath.digamma(-1e-17)
...
mpf('99999999999999992.26854192963627518392318558407754739227103018461090688982779538788336487955445392962043')
```

The old and new code should do the same thing on this code path; in the old version:

- `nz` is set to `pi / tan(pi *x)`, `x` is changed to `1 - x`
- `y` is initialized to `0`
- Formula for integers added to `y`; jump to `done`
- In `done` subtract `nz`

In the new version:

- `y` is initialized to `0`
- `pi / tan(pi * x)` is subtracted right away, `x` is changed to `1 - x`
- Formula for integers added to `y`; return.

And we don't lose precision because if `1 - x == 1`, then the function returns `-pi * cot(pi * x) - gamma`. Now `-pi * cot(pi * x) = -1/x + O(x)`, so from DLMF 5.7.6:

http://dlmf.nist.gov/5.7#E6

we're correct up to an `O(x)` correction term. But a correction of that size when won't show up when `1  - x == 1` because of the `O(1/x)` term.

A bit sneaky that it all works though...